### PR TITLE
wrap lesson update in a database transaction to avoid partial updates

### PR DESCRIPTION
**Pull request by davidsbailey**
Addresses this high-priority fix from the cplat internal bug bash doc:

> when saving failed, a subset of my changes persisted. This will probably be solved by having better error handling in general, but it’s hard to figure out what saved correctly and what didn’t

The solution is to wrap all DB activity within the update method in a database transaction.

This PR also does some cleanup to return 4xx errors instead of 5xx when certain errors occur.

## Testing story

New unit test verifying that partial updates are prevented.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by davidsbailey** [comment]
If you make an error by following the repro steps in https://codedotorg.atlassian.net/browse/PLAT-418, you'll now get an error page that looks like this:
Screen Shot 2020-10-26 at 7 54 31 AM

previously you would get a stack trace. 

this seems like the best we can do until we start using AJAX instead of form post to hit the server, after which we should be able to at least show these kinds of errors to the editor without making them navigate away and lose any updates.

File: dashboard/app/controllers/lessons_controller.rb [hunk 1]
```diff
@@ -43,13 +43,17 @@ def edit
 
   # PATCH/PUT /lessons/1
   def update
-    resources = (lesson_params['resources'] || []).map {|key| Resource.find_by_key(key)}
-    @lesson.resources = resources.compact
-    @lesson.update!(lesson_params.except(:resources, :objectives))
-    @lesson.update_activities(JSON.parse(params[:activities])) if params[:activities]
-    @lesson.update_objectives(JSON.parse(params[:objectives])) if params[:objectives]
+    ActiveRecord::Base.transaction do
+      resources = (lesson_params['resources'] || []).map {|key| Resource.find_by_key(key)}
```
Comment (bethanyaconnor) [inline dashboard/app/controllers/lessons_controller.rb]: Should these non-update statements be outside of the transaction or does it not matter?

**Review by bethanyaconnor — APPROVED** [review]
Thanks for doing this!

File: dashboard/app/controllers/lessons_controller.rb (hunk 1 above)
Comment (davidsbailey) [reply dashboard/app/controllers/lessons_controller.rb]: yes, probably best to do as little as possible inside the transaction!

**Pull request merged**